Fix minikube load on windows (#20529)

james-world opened the pull request:
fixes #20529

`getWindowsVolumeNameCmd` uses the deprecated wmic.exe, this breaks `minikube image load` on newer installs of windows. Whilst wmic.exe can be added, it will be removed from Windows 11 soon.

See [here](https://learn.microsoft.com/en-us/windows/win32/wmisdk/wmic) for details.

This fix replaces the use of `wmic.exe` with a call to the Powershell equivalent.

<!-- 🎉 Thank you for contributing to minikube! 🎉 Here are some hints to get your PR merged faster:

1. Your PR title will be included in the release notes, choose it carefully
2. If the PR fixes an issue, add "fixes #<issue number>" to the description.
3. If the PR is a user interface change, please include a "before" and "after" example.
4. If the PR is a large design change, please include an enhancement proposal:
https://github.com/kubernetes/minikube/tree/master/enhancements
-->

james-world commented on pkg/minikube/localpath/localpath.go:
```diff
@@ -190,25 +191,20 @@ func replaceWinDriveLetterToVolumeName(s string) (string, error) {
 
 func getWindowsVolumeNameCmd(d string) (string, error) {
-	cmd := exec.Command("wmic", "volume", "where", "DriveLetter = '"+d+":'", "get", "DeviceID")
+	psCommand := `Get-CimInstance -ClassName Win32_Volume -Filter "DriveLetter = '` + d + `:'" | Select-Object -ExpandProperty DeviceID`
```
Hi @medyagh,

Addressing your points in turn:

### 1. Output before and after

Before (on any windows system where wmic.exe is not in path, whether run from Git Bash / MingW64 / Cmd / Powershell):

```
.\minikube.exe image load employee-api:latest

❌  Exiting due to GUEST_IMAGE_LOAD: Failed to load image: save to dir: caching images: caching image "C:\\Users\\james.world\\.minikube\\cache\\images\\amd64\\employee-api_latest": getting destination path: parsing docker archive dst ref: replace a Win drive letter to a volume name: exec: "wmic": executable file not found in %PATH%

╭───────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                               │
│    😿  If the above advice does not help, please let us know:                                                 │
│    👉  https://github.com/kubernetes/minikube/issues/new/choose                                               │
│                                                                                                               │
│    Please run `minikube logs --file=logs.txt` and attach logs.txt to the GitHub issue.                        │
│    Please also attach the following file to the GitHub issue:                                                 │
│    - C:\Users\JAMES~1.WOR\AppData\Local\Temp\minikube_image_1b5c8c33fb94831b9d273a970d695f2957a83a6b_0.log    │
│                                                                                                               │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

Output after (again, on any windows system where wmic.exe is not in path, whether run from Git Bash / MingW64 / Cmd / Powershell):

_There is no output!_ , as with the existing implementation the command actually doesn't output anything when successful. This is a helper method executed by upstream functionality for image load - and that doesn't output anything on success bar the return code.

### 2. Adding fallback behaviour

As for adding the failover - on the surface it sounds like a good idea. I can do it if you insist, however consider that as far as I know the Get-CimInstance was first added in Windows Powershell 3.0 and has been included in windows by default since Windows 8 and Windows Server 2012. Do we really need this fallback, do you think?

### 3. Would this work in "git bash in windows" or only in powershell ?

It works regardless of the windows shell from which minikube is executed - the go code is launching the Powershell process itself; it should always be in path. I've certainly tested from Git Bash, MingW64, Powershell and Cmd and all are fine.

### 4. I think this is an opportunity to add unit test here.

What would a unit test do beyond what's already in localpath_test.go? This feels more like integration testing territory to me? Because in this case, the function is immediately calling out to powershell's Get-CimInstance command and therefore is very specific to Windows and the Powershell tooling.

Honestly, I have very little go experience. It's been novel for me to build minikube on Windows; the docs aren't very informative for Windows platform builds. Because this is a drop in replacement for existing functionality, I had hoped the existing test coverage would be fine. If not, do you have a good example you can point to for an existing integration test that is Windows specific that would serve as a good template for how to go about this idiomatically?

james-world replied:
On further consideration @medyagh, it's pretty trivial to add the fallback, so I have.

gabrielgbs97 replied:
I believe the older method should not be used as a fallback. WMIC **has been deprecated** since the release of Windows 10 RTM, which was 10 years ago, and it is not included in the default installation of Windows 11 23H2. Therefore, it would not have been implemented using wmic.exe around 2016 - 2019 (?)

bobsira replied:
https://github.com/minikube-machine/machine/blob/main/drivers/hyperv/powershell.go 

This currently exist on how we discover Powershell and execute PS commands, you can borrow from this. Your change looks alright to me.

james-world commented:
OK, sorry for the hiatus - been busy at work. @medyagh  I've removed the fallback as discussed and added code to robustly discover Powershell.exe, using the cited approaches.

medyagh commented on pkg/minikube/localpath/localpath.go:
```diff
@@ -189,27 +191,55 @@ func replaceWinDriveLetterToVolumeName(s string) (string, error) {
+	return "", fmt.Errorf("PowerShell not found in PATH or common locations")
+}
+
+func getWindowsVolumeNameCmd(d string) (string, error) {
```
add comment here pleae

bobsira replied:
The Get-CimInstance command works fine on my Windows machine!

bobsira reviewed (commented):
Tested this on Windows machine, works fine! Approved @medyagh 

Had a successful run with the changes,
`C:\dev\minikube\out>minikube.exe image load nginx:latest --alsologtostderr                                                                                                                                                                           I0714 02:30:21.918732   24816 out.go:345] Setting OutFile to fd 96 ...                                                                                                                                                                               I0714 02:30:21.949424   24816 out.go:392] TERM=,COLORTERM=, which probably does not support color                                                                                                                                                    I0714 02:30:21.949424   24816 out.go:358] Setting ErrFile to fd 100...                                                                                                                                                                               I0714 02:30:21.950091   24816 out.go:392] TERM=,COLORTERM=, which probably does not support color                                                                                                                                                    I0714 02:30:21.969639   24816 config.go:182] Loaded profile config "minikube": Driver=hyperv, ContainerRuntime=docker,
KubernetesVersion=v1.33.2                                                                                                     

=== some of the logs removed to have this fit in the comment
                                                                                                   
I0714 02:30:36.438734   24816 cache_images.go:123] Successfully loaded all cached images                                                                                                                                                             I0714 02:30:36.438734   24816 cache_images.go:92] duration metric: took 5.9631078s to LoadCachedImages                                                                                                                                               I0714 02:30:36.439366   24816 cache_images.go:262] succeeded pushing to: minikube  `

medyagh commented:
thank you @james-world

james-world commented:
@medyagh  I've started to see this kind of error at work; it's docker.io rate-limiting. Definitely nothing to do with this PR.

The rate limiting on anonymous logins is by IP - see [here](https://docs.docker.com/docker-hub/usage/).

We've addressed it by using pull-through caches (we use ACR for this: https://learn.microsoft.com/en-us/azure/container-registry/artifact-cache-overview)

Guess minikube would need to cut a deal with docker.io, or do something similar.
